### DISCUSSION TITLE
Best way to create a Vault to put Tokens in it? (others call it Pool)

### DISCUSSION THREAD

**Discussion by NaDasai** [aptos-labs/aptos-developer-discussions#416] [Questions] [answered]
### Discord user ID

ibuidl

### Describe your question in detail.

I want to create a Vault/Pool where when the users pays (let's say in APT) for a service, the Tokens are added to the Vault/Pool.
After that the one that provided the service would go and claim those tokens.
I tried with this, but I'm not sure it's the best way to do it:
```
    struct Pool has key {
        coins : coin::Coin<AptosCoin>,
    }
```
    then
```
        let pool = borrow_global_mut<Pool>(@module_addr);
        let coin = coin::extract(&mut pool.coins,amount);
```

### Which operating system are you using?

macOS

### Which SDK or tool are you using? (if any)

Rust SDK

**Comment by kshitijc1506** [comment, marked as answer]
You're on the right track with Coin. Here's a full [example](https://explorer.aptoslabs.com/account/0x02bf331b3010eccd0cc255e7338220d9c5c5aff1c363e230c80a7a08da608860/modules/code/ae_rewards?network=mainnet) that supports both Coin and FA.

**Reply by NaDasai** [reply]
1/ About this part:
```
    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
    struct FARewards has key {
        rewards: SmartTable<String, FALockBox>,
    }
```
Can you explain what's the _resource_group_member_ about?
And the SmartTable is it the equivalent of KeyValueStore in Rust? And when you use a "smart" table, how is it different from normal one?

2/  I guess this ise like the one I used:
```
    struct CoinRewardPool<phantom CoinType> has key {
        project: String,
        coins: Coin<CoinType>,
        distribution_amount: u64,
        claimed: vector<address>
    }
```
Why is _<phantom CoinType>_ added?

3/ So this is the FA (what's FA btw?) you mentionned:
```
    struct FALockBox has copy, drop, store {
        pool: Object<FungibleStore>,
        distribution_amount: u64,
        claimed: vector<address>
    }
```
When do you use a _FungibleStore_ and why do we use _Object_?

**Reply by kshitijc1506** [reply]
(1) that is the [resource group](https://github.com/aptos-foundation/AIPs/blob/main/aips/aip-9.md) annotation, it treats those resources as a single resource in global storage / stores them in a single storage slot. 

SmartTable is designed to be more flexible and gas efficient than a regular table. read more [here](https://x.com/Greg_Nazario/status/1760809633399366013). 

(2) not sure what you mean by this question

(3) FA is the [Fungible Asset Standard](https://aptos.dev/en/build/smart-contracts/fungible-asset). it's the newer standard for fungible tokens and uses the aptos object model as its building blocks.

**Reply by NaDasai** [reply]
(1) Is there a reason why we should store in a single slot?
(2) sorry for got to copy past: why is _<phantom CoinType>_  used here?
(3) So it's better recommend for the Vault/pool? If yes, why?

**Reply by kshitijc1506** [reply]
for (1) you can read the AIP I linked, has all the details :)

for (2) and (3), since we're in the middle of migration, it makes sense to support both. read more about the migration [here](https://aptos.dev/en/build/smart-contracts/fungible-asset#migration-from-coin-to-the-fungible-asset-standard).
